library/math/elementary.lisp

```diff
@@ -0,0 +1,373 @@
+;;;; elementary.lisp
+;;;;
+;;;; Elementary/algebriac functions and trancsendental numbers
+
+(coalton-library/utils::defstdlib-package #:coalton-library/math/elementary
+    (:use
+     #:coalton
+     #:coalton-library/builtin
+     #:coalton-library/classes
+     #:coalton-library/functions
+     #:coalton-library/utils
+     #:coalton-library/math/arith
+     #:coalton-library/math/real
+     #:coalton-library/math/complex)
+  (:local-nicknames (#:ff #:float-features))
+  (:export
+   #:Trigonometric
+   #:sin #:cos #:tan
+   #:sinh #:cosh #:tanh
+   #:asin #:acos #:atan
+   #:asinh #:acosh #:atanh
+   #:sincos
+   #:Radical
+   #:sqrt
+   #:nth-root
+   #:Polar
+   #:phase
+   #:magnitude
+   #:cis
+   #:Exponentiable
+   #:log #:pow #:exp #:ln
+   #:Elementary
+   #:pi #:ee
+   #:atan2))
+
+#+coalton-release
+(cl:declaim #.coalton-impl:*coalton-optimize-library*)
+
+(in-package #:coalton-library/math/elementary)
+
+(coalton-toplevel
+  (define-class (Trigonometric :a)
+    "Standard circular functions and their inverses."
+    (sin   (:a -> :a))
+    (cos   (:a -> :a))
+    (tan   (:a -> :a))
+    (asin  (:a -> :a))
+    (acos  (:a -> :a))
+    (atan  (:a -> :a)))
+
+  (declare sincos (Trigonometric :a => :a -> (Tuple :a :a)))
+  (define (sincos x)
+    (Tuple (sin x) (cos x)))
+
+  (define-class (Exponentiable :a)
+    "Exponential maps obeying:
+
+    - (* (exp x) (exp y)) = (exp (+ x y))
+    - (exp (ln x)) = x = (ln (exp x))
+    - (log b x) = (/ (ln x) (ln b))
+    - (pow x y) = (exp (* y (ln x)))
+"
+    (exp (:a -> :a))
+    (pow (:a -> :a -> :a))
+    (ln  (:a -> :a))
+    (log  (:a -> :a -> :a)))
+
+  (define-class (Radical :a)
+    "Obeys, (^ (sqrt x) 2) = x and (^^ (nth-root n x) n) = x"
+    (nth-root (Integer -> :a -> :a))
+    (sqrt (:a -> :a)))
+
+  (define-class ((Complex :a) (Num :a) => Polar :a)
+    "For a complex number z = (complex x y)
+    z = (* (magnitude z) (exp (* ii (phase z))))
```

Comment:
you might want to format these as markdown

Reply:
I agree, but I am going to wait until after #612 so I can see the output.